python/ql/src/experimental/Security/CWE-287-ConstantSecretKey/ConstantSecretKey.ql

```diff
@@ -0,0 +1,219 @@
+   * Also Sources can be the default value of getenv or similar methods
+   * in a case that no value is assigned to Desired SECRET_KEY environment variable
+   */
+  predicate isSource(DataFlow::Node source) {
+    (
+      source.asExpr().isConstant()
+      or
+      exists(API::Node cn |
+        cn =
+          [
+            API::moduleImport("configparser")
+                .getMember(["ConfigParser", "RawConfigParser"])
+                .getReturn(),
+            // legacy API https://docs.python.org/3/library/configparser.html#legacy-api-examples
+            API::moduleImport("configparser")
+                .getMember(["ConfigParser", "RawConfigParser"])
+                .getReturn()
+                .getMember("get")
+                .getReturn()
+          ] and
+        source = cn.asSource()
+      )
+      or
+      exists(API::CallNode cn |
+        cn =
+          [
+            API::moduleImport("os").getMember("getenv").getACall(),
+            API::moduleImport("os").getMember("environ").getMember("get").getACall()
+          ] and
+        (
+          // this can be ideal if we assume that best security practice is that
+          // we don't get SECRET_KEY from env and we always assign a secure generated random string to it
+          cn.getNumArgument() = 1
+          or
+          cn.getNumArgument() = 2 and
+          DataFlow::localFlow(any(DataFlow::Node n | n.asExpr().isConstant()), cn.getArg(1))
+        ) and
+        source.asExpr() = cn.asExpr()
+      )
+      or
+      exists(DataFlow::LocalSourceNode lsn |
+        lsn = API::moduleImport("os").getMember("environ").getASubscript().asSource() and
+        source.asExpr() = lsn.asExpr()
+      )
+    ) and
+    not source.getScope().getLocation().getFile().inStdlib()
+  }
+
+  /**
+   * Sinks are one of the following kinds, some of them are directly connected to a flask Instance like
+   * ```python
+   *    app.config['SECRET_KEY'] = 'CHANGEME1'
+   *    app.secret_key = 'CHANGEME2'
+   *    app.config.update(SECRET_KEY="CHANGEME3")
+   *    app.config.from_mapping(SECRET_KEY="CHANGEME4")
+   * ```
+   * other Sinks are SECRET_KEY Constants Variables that are defined in seperate files or a class in those files like:
+   * ```python
+   *    app.config.from_pyfile("config.py")
+   *    app.config.from_object('config.Config')
+   *```
+   * we find these files with `FromObjectFileName` DataFlow Configuration
+   * note that "JWT_SECRET_KEY" is same as "SECRET_KEY" but it is belong to popular flask-jwt-extended library
+   */
+  predicate isSink(DataFlow::Node sink) {
+    (
+      exists(API::Node n |
+        n = flaskInstance() and
+        flask_sessionSanitizer(n.getReturn().asSource())
+      |
+        sink =
+          [
+            n.getReturn().getAMember().getSubscript(["SECRET_KEY", "JWT_SECRET_KEY"]).asSink(),
+            n.getReturn().getMember(["SECRET_KEY", "JWT_SECRET_KEY"]).asSink(),
```

Comment:
As per https://flask.palletsprojects.com/en/2.3.x/api/#flask.Flask.secret_key

```suggestion
            n.getReturn().getMember(["secret_key", "JWT_SECRET_KEY"]).asSink(),
```

Although I'm not sure this will cover the value assigned in `app.secret_key = <bad>`, I think you would be better off using `DataFlow::AttrWrite attr`, and use the _value_ that is being assigned as the sink :blush: